pkg/registry/node/etcd/etcd.go

```diff
@@ -131,17 +132,47 @@ func (r *REST) getKubeletPort(ctx api.Context, nodeName string) (int, error) {
 	return int(node.Status.DaemonEndpoints.KubeletEndpoint.Port), nil
 }
 
-func (c *REST) GetConnectionInfo(ctx api.Context, nodeName string) (string, uint, http.RoundTripper, error) {
-	scheme, port, transport, err := c.connection.GetConnectionInfo(ctx, nodeName)
+func getNodeAddress(node *api.Node) string {
+	nodeAddresses := node.Status.Addresses
+	for _, address := range nodeAddresses {
+		if address.Type == api.NodeHostName {
+			return address.Address
+		}
+	}
+	nodeName := node.Name
+	glog.Warningf("Failed to retrieve Hostname for Node %s, falling back to NodeName", nodeName)
+	return nodeName
+}
+
+func (r *REST) getKubeletHost(nodeName string) (string, error) {
+	obj, err := r.Get(api.NewDefaultContext(), nodeName)
+	if err != nil {
+		return "", err
+	}
+	node, ok := obj.(*api.Node)
+	if !ok {
+		return "", fmt.Errorf("Unexpected object type: %#v", node)
+	}
+
+	return getNodeAddress(node), nil
+}
+
+func (c *REST) GetConnectionInfo(ctx api.Context, nodeName string) (*client.ConnectionInfo, error) {
+	hostname, err := c.getKubeletHost(nodeName)
```

Comment:
this is forcing two lookups of the node object from etcd... (once for the host, once for the port)... just fetch once